### PR TITLE
UpdateChecker

### PR DESCRIPTION
**Pull request by IhateTrains**
This is what appears if a tag with higher version is found on startup:
image

A button:
image

With `enableUpdateChecker = false`, the messages and the button don't appear.
With `checkForUpdatesOnStartup = false`, the check is only performed on button click.

Example configuration:

> enableUpdateChecker = true
checkForUpdatesOnStartup = true
converterTagsGithubUrl = "https://api.github.com/repos/ParadoxGameConverters/CK2ToEU4/tags"
converterReleaseForumThread = "https://forum.paradoxplaza.com/forum/threads/ck2-to-eu4-converter-release-thread.1377860/"

### CODE REVIEW

**Comment by IhateTrains** [comment]
Ideas for improvement?

File: Fronter/Source/Frames/MainFrame.cpp [hunk 1]
```diff
@@ -99,6 +101,17 @@ void MainFrame::OnSupportUs(wxCommandEvent& event)
 	wxLaunchDefaultBrowser("https://www.patreon.com/ParadoxGameConverters");
 }
 
+void MainFrame::OnCheckForUpdates(wxCommandEvent& event)
+{
+	if (isUpdateAvailable(configuration->getConverterFolder() + "/configurables/version.txt", configuration->getConverterTagsGithubUrl()))
```
Comment (Zemurin) [inline Fronter/Source/Frames/MainFrame.cpp:106]: does this work if there's no getConverterFolder()?
on linux I bet it'll break.

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+bool newerVersion(const std::string& currentVersion, const std::set<std::string>& versions)
+{
+	auto newestVersion = currentVersion;
+	auto changed = false;
+	for (const auto& version : versions)
+	{
+		const auto newestDotPos = newestVersion.find('.');
+		const auto versionDotPos = version.find('.');
+		
+		if (std::stoi(version) > std::stoi(newestVersion))
```
Comment (Zemurin) [inline Fronter/Source/UpdateChecker/UpdateChecker.cpp:15]: please try/catch around all stois.

**Review by Zemurin — CHANGES_REQUESTED** [review]

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+std::set<std::string> getAllNumberedVersions(std::string& json)
+{
+	std::set<std::string> versions;
+	size_t pos;
+	size_t nameStart;
+	size_t nameEnd = 0;
+	do
+	{
+		pos = json.find("\"name\"", nameEnd);
```
Comment (Idhrendur) [inline Fronter/Source/UpdateChecker/UpdateChecker.cpp:87]: I'm torn. On the one hand, that's not how you're supposed to do json, and it's likely really fragile. On the other, it saved us a json library.

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp (hunk 2 above)
Comment (Idhrendur) [inline Fronter/Source/UpdateChecker/UpdateChecker.cpp:15]: I think you meant to extract the left sides here.

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+			changed = true;
+			continue;
+		}
+		if (std::stoi(version) < std::stoi(newestVersion))
+			continue;
+
+		// if parts before dot are equal, compare the rest
+		auto newestRSide = newestDotPos == std::string::npos ? "" : newestVersion.substr(newestDotPos + 1);
+		auto versionRSide = versionDotPos == std::string::npos ? "" : version.substr(versionDotPos + 1);
+		if (version.size() > newestVersion.size())
```
Comment (Idhrendur) [inline Fronter/Source/UpdateChecker/UpdateChecker.cpp:27]: You should be able to replace these loops with judicious use of std::transform (see the first example [here](https://en.cppreference.com/w/cpp/algorithm/transform)) and std::lexicographical_compare

Comment (Idhrendur) [reply Fronter/Source/UpdateChecker/UpdateChecker.cpp:27]: It might even invalidate the need to compare both parts separately.

**Review by Idhrendur — CHANGES_REQUESTED** [review]
More straightforward than I expected, but I hadn't ever really put thought to version checking.

However, it may also need a special case for prereleases. We use the convention of adding "-prerelease" to the version number, which would make a prerelease always seem newer than the equivalent full release. Dev versions are kind of an edge case to begin with, though. Maybe this is good enough for main versions, and some followup work can address guiding people to the latest dev release if they're using a dev release.

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+static char errorBuffer[CURL_ERROR_SIZE];
+static std::string buffer;
+
+//
+//  libcurl write callback function
+//
+
+static int writer(char* data, size_t size, size_t nmemb, std::string* writerData)
+{
+	if (writerData == NULL)
```
[the remark this answers is not in the document]
Comment (IhateTrains) [reply Fronter/Source/UpdateChecker/UpdateChecker.cpp:118]: Not sure if I can replace with `nullptr` as the NULL comes from CURL, but I will replace with `if (!writerData)`.

File: Fronter/Source/Frames/MainFrame.cpp (hunk 1 above)
Comment (IhateTrains) [reply Fronter/Source/Frames/MainFrame.cpp:106]: Made the function break when `getConverterFolder()` returns an empty string.

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp (hunk 2 above)
Comment (IhateTrains) [reply Fronter/Source/UpdateChecker/UpdateChecker.cpp:15]: Doesn't `stoi` do that for us? For example `stoi("1.5F")` should return 1.

File: Copy_Files.bat [hunk 6]
```diff
@@ -0,0 +1,18 @@
+echo on
```
Comment (Idhrendur) [inline Copy_Files.bat:1]: Wouldn't just adding these files to the project in Visual Studio allow you to remove the triplicate copy of Copy_Files?

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp (hunk 2 above)
Comment (Idhrendur) [reply Fronter/Source/UpdateChecker/UpdateChecker.cpp:15]: Hmm, perhaps technically true, but take my comment as an indication that it's unclear to people reading the code.

File: Copy_Files.bat (hunk 6 above)
Comment (Zemurin) [reply Copy_Files.bat:1]: I already did that. There's no copy_files.bat in current version. What you're looking at is some new code merged with all the reverts back to some ancient fronter version.

**Comment by Zemurin** [comment]
Please make a copy of your fork somewhere, make a clean sync with current, kill this PR, overwrite what's changed, and make a fresh PR so I can look at stuff that's actually changed. I removed those copy bats, for example, on Oct 7.

You reverted a lot of code, and you're reapplying a ton of existing changes.

**Comment by IhateTrains** [comment]
> Please make a copy of your fork somewhere, make a clean sync with current, kill this PR, overwrite what's changed, and make a fresh PR so I can look at stuff that's actually changed. I removed those copy bats, for example, on Oct 7.
> 
> You reverted a lot of code, and you're reapplying a ton of existing changes.

No need to, I removed the `.bat`s and based the `.vcxproj` on upstream/master. I don't see code reversions, can you recheck the diff?
The change from `loglevel` to `logLevel` was intentional, because that didn't match camelCase.

File: log.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+2020-12-11 19:15:44  [WARNING] Configuration/fronter-configuration.txt not found!
```
Comment (Idhrendur) [inline log.txt:1]: log.txt shouldn't be checked in.

File: Fronter/Source/UpdateChecker/UpdateChecker.cpp (hunk 3 above)
Comment (IhateTrains) [reply Fronter/Source/UpdateChecker/UpdateChecker.cpp:87]: We could use this: https://github.com/nlohmann/json

File: Copy_Files.bat (hunk 6 above)
Comment (IhateTrains) [reply Copy_Files.bat]: Solved ✅

**Pull request closed without merging**